File: factory/scraper_factory.py
import logging
from typing import Optional, Type
from scrapers.base_scraper import BaseScraper
from scrapers.cnn_scraper import CNNScraper
from scrapers.fox_scraper import FoxScraper
from scrapers.theatlantic_scraper import TheAtlanticScraper
from scrapers.theverge_scraper import TheVergeScraper
from config import SUPPORTED_SITES

logger = logging.getLogger(__name__)


class ScraperFactory:
    _scraper_map = {
        "CNN": CNNScraper,
        "Fox News": FoxScraper,
        "The Atlantic": TheAtlanticScraper,
        "The Verge": TheVergeScraper
    }
    
    @classmethod
    def create_scraper(cls, url: str) -> Optional[BaseScraper]:
        scraper_class = cls.get_scraper_class(url)
        if scraper_class:
            return scraper_class()
        return None
    
    @classmethod
    def get_scraper_class(cls, url: str) -> Optional[Type[BaseScraper]]:
        if not url:
            return None
        
        url_lower = url.lower().strip()
        
        for site_name, domains in SUPPORTED_SITES.items():
            for domain in domains:
                if domain in url_lower:
                    scraper_class = cls._scraper_map.get(site_name)
                    if scraper_class:
                        logger.info(f"Matched URL '{url}' to {site_name} scraper")
                        return scraper_class
        
        logger.warning(f"No scraper found for URL: {url}")
        return None
    
    @classmethod
    def is_supported_url(cls, url: str) -> bool:
        return cls.get_scraper_class(url) is not None
    
    @classmethod
    def get_supported_sites_list(cls) -> str:
        return ", ".join(SUPPORTED_SITES.keys())

File: factory/scraper-factory.js
import { logger, SUPPORTED_SITES } from '../config.js';
import { CNNScraper } from '../class/cnn-scraper.js';
import { FoxScraper } from '../class/fox-scraper.js';
import { TheAtlanticScraper } from '../class/theatlantic-scraper.js';
import { TheVergeScraper } from '../class/theverge-scraper.js';
import { VoxScraper } from '../class/vox-scraper.js';
import { APNewsScraper } from '../class/apnews-scraper.js';
import { NationalReviewScraper } from '../class/nationalreview-scraper.js';
import { AxiosScraper } from '../class/axios-scraper.js';
import { PennLiveScraper } from '../class/pennlive-scraper.js';
import { TheGuardianScraper } from '../class/theguardian-scraper.js';
import { TechCrunchScraper } from '../class/techcrunch-scraper.js';
import { QuartzScraper } from '../class/quartz-scraper.js';

export class ScraperFactory {
    static _scraperMap = {
        "CNN": CNNScraper,
        "Fox News": FoxScraper,
        "The Atlantic": TheAtlanticScraper,
        "The Verge": TheVergeScraper,
        "Vox": VoxScraper,
        "AP News": APNewsScraper,
        "National Review": NationalReviewScraper,
        "Axios": AxiosScraper,
        "PennLive": PennLiveScraper,
        "The Guardian": TheGuardianScraper,
        "TechCrunch": TechCrunchScraper,
        "Quartz": QuartzScraper
    };

    static createScraper(url) {
        const ScraperClass = this.getScraperClass(url);
        if (ScraperClass) {
            return new ScraperClass();
        }
        return null;
    }

    static getScraperClass(url) {
        if (!url) {
            return null;
        }
        
        const urlLower = url.toLowerCase().trim();
        
        for (const [siteName, domains] of Object.entries(SUPPORTED_SITES)) {
            for (const domain of domains) {
                if (urlLower.includes(domain)) {
                    const ScraperClass = this._scraperMap[siteName];
                    if (ScraperClass) {
                        logger.info(`Matched URL '${url}' to ${siteName} scraper`);
                        return ScraperClass;
                    }
                }
            }
        }
        
        logger.warn(`No scraper found for URL: ${url}`);
        return null;
    }

    static isSupportedUrl(url) {
        return this.getScraperClass(url) !== null;
    }

    static getSupportedSitesList() {
        return Object.keys(SUPPORTED_SITES).join(", ");
    }
}

